File: ui/trading/src/graphql/mt5.ts
import { gql } from '@apollo/client';

export const CONNECT_MT5_MUTATION = gql`
  mutation ConnectMt5($account: MT5AccountInput!) {
    connectMt5(account: $account) {
      id
      accountLogin
      server
      isConnected
      lastPing
      errorMessage
      createdAt
    }
  }
`;

export const MT5_ACCOUNT_INFO_QUERY = gql`
  query Mt5AccountInfo {
    mt5AccountInfo {
      login
      server
      name
      company
      currency
      balance
      equity
      margin
      marginFree
      leverage
    }
  }
`;

export const MT5_POSITIONS_LIVE_QUERY = gql`
  query Mt5PositionsLive {
    mt5PositionsLive {
      ticket
      symbol
      volume
      type
      priceOpen
      priceCurrent
      profit
      magic
      sl
      tp
    }
  }
`;

export const MT5_ORDERS_QUERY = gql`
  query Mt5Orders {
    mt5Orders {
      ticket
      symbol
      volumeCurrent
      type
      priceOpen
      sl
      tp
      magic
      state
    }
  }
`;

export const MT5_TRADING_HISTORY_QUERY = gql`
  query TradingHistory {
    tradingHistory {
      id
      symbol
      volume
      type
      price
      profit
      commission
      swap
      ticketId
      magic
      openTime
      closeTime
    }
  }
`;

export const PLACE_ORDER_MUTATION = gql`
  mutation PlaceOrder($order: OrderInput!) {
    placeOrder(order: $order) {
      id
      symbol
      volume
      type
      price
      sl
      tp
      status
      ticketId
      magic
      createdAt
      filledAt
    }
  }
`;

export const CLOSE_POSITION_MUTATION = gql`
  mutation ClosePosition($positionId: String!) {
    closePosition(positionId: $positionId)
  }
`;

File: trading/app/models/mt5.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from beanie import Document, Indexed
from pymongo import IndexModel
import json
import base64
from cryptography.fernet import Fernet


class MT5Account(BaseModel):
    """MT5 Account credentials"""
    login: str
    password: str
    server: str = Field(default="MetaQuotes-Demo")  # Default MT5 demo server
    path: str = Field(default="")  # MT5 terminal path


class Position(Document):
    """Trading Position Model"""
    symbol: str = Indexed()
    volume: float
    type: str  # BUY/SELL
    price_open: float
    price_current: float = 0.0
    profit: float = 0.0
    user_id: str = Indexed()
    ticket_id: Optional[str] = None
    magic: Optional[int] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "positions"
        

class Order(Document):
    """Trading Order Model"""
    symbol: str = Indexed()
    volume: float
    type: str  # BUY/SELL
    price: float
    sl: Optional[float] = None
    tp: Optional[float] = None
    status: str = "PENDING"  # PENDING/FILLED/CANCELLED
    user_id: str = Indexed()
    ticket_id: Optional[str] = None
    magic: Optional[int] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    filled_at: Optional[datetime] = None
    
    class Settings:
        name = "orders"


class Trade(Document):
    """Trade History Model"""
    symbol: str = Indexed()
    volume: float
    type: str  # BUY/SELL
    price: float
    profit: float
    commission: float
    swap: float
    user_id: str = Indexed()
    ticket_id: Optional[str] = None
    magic: Optional[int] = None
    open_time: datetime
    close_time: Optional[datetime] = None
    
    class Settings:
        name = "trades"


class MT5Connection(Document):
    """MT5 Connection Status"""
    user_id: str = Indexed()
    account_login: str
    server: str
    is_connected: bool = False
    last_ping: Optional[datetime] = None
    error_message: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "mt5_connections"
        use_state_management = True
        indexes = [
            IndexModel([("server", 1), ("account_login", 1)], unique=True),
        ]
